agentendpoint/config_task.go

```diff
@@ -0,0 +1,462 @@
+			ApplyConfigTaskOutput: &agentendpointpb.ApplyConfigTaskOutput{State: state, ConfigAssignmentResults: c.results},
+		},
+	}
+	if err := c.client.reportTaskComplete(ctx, req); err != nil {
+		return fmt.Errorf("error reporting completed state: %v", err)
+	}
+	return nil
+}
+
+func (c *configTask) handleErrorState(ctx context.Context, msg string, err error) error {
+	if err == errServerCancel {
+		clog.Infof(ctx, "Canceling config run: %v", errServerCancel)
+		return c.reportCompletedState(ctx, errServerCancel.Error(), agentendpointpb.ApplyConfigTaskOutput_CANCELLED)
+	}
+	msg = fmt.Sprintf("%s: %v", msg, err)
+	clog.Errorf(ctx, msg)
+	return c.reportCompletedState(ctx, msg, agentendpointpb.ApplyConfigTaskOutput_FAILED)
+}
+
+func (c *configTask) reportContinuingState(ctx context.Context, configState agentendpointpb.ApplyConfigTaskProgress_State) error {
+	req := &agentendpointpb.ReportTaskProgressRequest{
+		TaskId:   c.TaskID,
+		TaskType: agentendpointpb.TaskType_APPLY_CONFIG_TASK,
+		Progress: &agentendpointpb.ReportTaskProgressRequest_ApplyConfigTaskProgress{
+			ApplyConfigTaskProgress: &agentendpointpb.ApplyConfigTaskProgress{State: configState},
+		},
+	}
+	res, err := c.client.reportTaskProgress(ctx, req)
+	if err != nil {
+		return fmt.Errorf("error reporting task progress %s: %v", configState, err)
+	}
+	if res.GetTaskDirective() == agentendpointpb.TaskDirective_STOP {
+		return errServerCancel
+	}
+	return nil
+}
+
+func (c *configTask) validation(ctx context.Context) {
+	// First populate validation results and internal assignment state.
+	c.assignments = map[string]*assignment{}
+	for i, a := range c.Task.GetConfig().GetConfigAssignments() {
+		asgnmnt := &assignment{policies: map[string]*policy{}}
+		c.assignments[a.GetConfigAssignment()] = asgnmnt
+		aResult := c.results.GetResults()[i]
+		for i, p := range a.GetPolicies() {
+			plcy := &policy{resources: map[string]*resource{}}
+			asgnmnt.policies[p.GetId()] = plcy
+			pResult := aResult.GetOsPolicyResults().GetResults()[i]
+			for i, r := range p.GetResources() {
+				plcy.resources[r.GetId()] = &resource{}
+				result := pResult.GetResourceResults().GetResults()[i]
+				result.GetExecutionSteps()[validationStepIndex] = &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_ExcecutionStep{
+					Step: &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_ExcecutionStep_Validation{
+						Validation: &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_Validation{
+							Outcome: agentendpointpb.ApplyConfigTaskOutput_ResourceResult_Validation_SKIPPED,
+						},
+					},
+				}
+			}
+		}
+	}
+
+	// Unmarshal all resources.
+	for i, a := range c.Task.GetConfig().GetConfigAssignments() {
+		ctx = clog.WithLabels(ctx, map[string]string{"config_assignment": a.GetConfigAssignment()})
+		asgnmnt := c.assignments[a.GetConfigAssignment()]
+		aResult := c.results.GetResults()[i]
+		for i, p := range a.GetPolicies() {
+			plcy := asgnmnt.policies[p.GetId()]
```

Comment:
do you have to update ctx to add policy id to the labels?
similar comment for resource id

Reply:
Just missed it for this function, added